First fixes for shellcheck

aragon999 opened the pull request:
So I worked a bit through the output of `shellcheck`. If you see anything problematic, or something you would like to solve in a different way, let me know.

The following two things are still open (I also wrote a short note in the code):
https://www.shellcheck.net/wiki/SC2076
https://www.shellcheck.net/wiki/SC2086

The `$makepkg_options` I would create as array and just add all the options.
About the other issue I am not really sure why this was done like that. I will look into that tomorrow.

pbrisbin commented on aurget:
```diff
@@ -158,10 +158,12 @@ sort_rpc() {
 }
 
+# TODO: SC2076: Don't quote right-hand side of =~, it'll match literally rather than as a regex.
 keep_sources() { [[ " $keep_source_packages " =~ " $1 " ]]; }
```
To really resolve this issue, `keep_source_packages` should be an array and we should be checking if `$1` is an element in it here. This has always been a bit of a hack. That said, I'm OK with the hack and would use `shellcheck disable=` to not report on this issue.

aragon999 replied:
Okay, I added a comment about the matching, the best I found on StackOverflow. 
Since I did not know what exactly happens there :-) is this okay?

pbrisbin commented on aurget:
```diff
@@ -193,16 +195,18 @@ buildpkg() {
   if $silent; then
     debug 'building silently...'
+    # TODO: SC2086: Double quote to prevent globbing and word splitting.
     $makepkg $makepkg_options "$@" --noconfirm &>/dev/null
```
I think we expect `makepkg` to always be a single word, so `"$makepkg"` should work here. I don't even know why we have this as configurable to be honest. Doe anyway use some alternative `makepkg`?

The `makepkg_options` variable should definitely be made an array, IMO. I can't remember if there's complexity around why it's not.

pbrisbin commented on aurget:
```diff
@@ -473,23 +477,26 @@ print_info() {
-  arguments=($(available_upgrades) "${arguments[@]}")
+
+  mapfile -t update_packages < <(available_upgrades)
+  arguments+=("${update_packages[@]}")
```
This actually changed behavior. Previously the result was `arguments = upgrades + arguments` but now it's `arguments = arguments + upgrades`. I don't know if I care, but I think I was trying to match what `pacman -Syu x y z` does in terms of processing available upgrades then command line arguments :man_shrugging:

aragon999 replied:
That is something, that should probably be changed again, I will think of a solution :-) 
It is probably the better thing to do, first upgrade the installed packages then install new ones.

aragon999 commented:
I added one test and modified the build test, such that the `makepkg` options are also checked. I think the changes are finished so far. Do you see something else which should be changed?

pbrisbin commented on aurget:
```diff
@@ -189,20 +192,20 @@ enter_build_directory() {
   if $silent; then
     debug 'building silently...'
-    $makepkg $makepkg_options "$@" --noconfirm &>/dev/null
+    $makepkg "${makepkg_options[@]}" "$@" --noconfirm &>/dev/null
```
I'm surprised ShellCheck doesn't complain about the un-quoted `$makepkg`, do you know why?

aragon999 replied:
I was surprised as well, that `$makepkg` does not need quotes (I initially thought that the error is only shown if the one with the `$makepkg_options` issue is resolved.

Should I add the quotes? Should I do that in another commit, such that it might be easier to revert this change in future? But I cannot find anything regarding this, nor do I know of any problems.

aragon999 replied:
Mh I looked into that, I think we should quote it.

The only disadvantage would be that it is not possible to put options, e.g. one cannot set `makepkg -d` as default, but that is something one probably do not want to do, since the `-d` belongs into the options. I cannot find any other difference.

pbrisbin commented on test/build.t:
```diff
@@ -2,14 +2,24 @@
 Building the package for aurget
 
-  $ aurget -Sb --builddir "$PWD" --noedit --nodeps --noconfirm --nodiscard aurget
+  $ aurget -Sb --builddir "$PWD" --noedit --nodeps --noconfirm --nodiscard --debug --nosign aurget
```
I think I would take the `--debug` back out of here. I can see it being useful to verify _this_ PR as you worked, but now that you're done, I think it's not worth the maintenance burden going forward. Assertions against something like debug logging are a bad idea because they're brittle: changes in them would have nothing to do with functionality but would cause the test to fail anyway.

WDYT?

aragon999 replied:
About the debug flag, you are right, that should probably be not in the tests, since it will cause to much trouble. 
It might be something useful though to see the `$makepkg_options` before `makepkg` is called, without having `debug` enabled. But that is something for the future I guess, I will remove the test commit. 

Do you think we should add another test where we only check the debug line of the `makepkg_options` and `(glob)` the rest?

pbrisbin replied:
> Do you think we should add another test where we only check the debug line of the makepkg_options and (glob) the rest?

Do we have any tests that attempt to pass `makepkg` options along? If we did, then I'd conser this covered already since the fixture system records commands with their options passed -- if that code regressed, we'd get a "no fixture" error. Not the most useful failure, but it is good regression coverage IMO.

aragon999 replied:
I think most of the builtin options are tested, but for example custom flags are not tested yet. Maybe it would be good to add one as I did in the test.

pbrisbin commented on test/upgrades.t:
```diff
@@ -12,3 +12,12 @@ specs with real network.)
   
   Proceed with installation? [Y/n]  (no-eol)
+
+Upgrades with some additional packages
```
:+1:

Nit, but can you put a line of whitespace after this, like the other examples do?

aragon999 commented:
Hey, I will address the comments and fix them in this PR, probably today :-)

pbrisbin commented:
Thanks!

aragon999 commented:
Thank you again for merging. 

One little question, when (or which features other) do you plan for the next release? Can I help with something?

pbrisbin commented:
> One little question, when (or which features other) do you plan for the next release? Can I help with something?

Releasing isn't a big deal, so I'm happy to do it whenever. Basically any time a PR lands that fixes a bug or adds a feature, I'll release shortly after. I probably won't release for this PR alone, since it's not user-facing.

Pushing along other open PRs or triaging and/or addressing open Issues would be my first preference for helping out. After that, I'm happy to review and hopefully release whatever you're interested in doing!

aragon999 commented:
> Releasing isn't a big deal, so I'm happy to do it whenever. Basically any time a PR lands that fixes a bug or adds a feature, I'll release shortly after. I probably won't release for this PR alone, since it's not user-facing.

I see. Yes I was not asking because of this PR, but the other one (#70) with the gawk issue. Since I currently have `aurget` not patched on one system, and notice it there every day :-)
It might probably be good, to release an update for this problem, see e.g. the following output:
```
$ aurget -Syu     
:: Starting AUR upgrade...
awk: cmd:4: Warning: regexp escape sequence `\"' is not a known regexp operator
:: Searching AUR...
awk: cmd:4: Warning: regexp escape sequence `\"' is not a known regexp operator

Targets (3): adminer-4.7.3-1 intellij-idea-ultimate-edition-2019.2.2-1 intellij-idea-ultimate-edition-jre-2019.2.2-1
```

It is somewhat harder to quickly scan the output.

pbrisbin commented:
Oh sorry. I don't need convincing, I *meant* to release that right away.
Will do today.

On Wed, Sep 11, 2019, 03:16 Max <notifications@github.com> wrote:

> Releasing isn't a big deal, so I'm happy to do it whenever. Basically any
> time a PR lands that fixes a bug or adds a feature, I'll release shortly
> after. I probably won't release for this PR alone, since it's not
> user-facing.
>
> I see. Yes I was not asking because of this PR, but the other one (#70
> <https://github.com/pbrisbin/aurget/pull/70>) with the gawk issue. Since
> I currently have aurget not patched on one system, and notice it there
> every day :-)
> It might probably be good, to release an update for this problem, see e.g.
> the following output:
>
> $ aurget -Syu
> :: Starting AUR upgrade...
> awk: cmd:4: Warning: regexp escape sequence `\"' is not a known regexp operator
> :: Searching AUR...
> awk: cmd:4: Warning: regexp escape sequence `\"' is not a known regexp operator
>
> Targets (3): adminer-4.7.3-1 intellij-idea-ultimate-edition-2019.2.2-1 intellij-idea-ultimate-edition-jre-2019.2.2-1
>
> It is somewhat harder to quickly scan the output.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/pbrisbin/aurget/pull/76?email_source=notifications&email_token=AAAMM7A4BBHDVIHFTLYJYATQJCLN7A5CNFSM4IUC7WN2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6NQV4I#issuecomment-530254577>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAMM7CJMMNNL5UXWG3EBD3QJCLN7ANCNFSM4IUC7WNQ>
> .
>
